core/src/main/scala/zio/config/ConfigSource.scala

```diff
@@ -16,28 +16,37 @@ import zio.Task
 
 import scala.annotation.tailrec
 
-final case class ConfigSource[K, V](
-  getConfigValue: Vector[K] => PropertyTree[K, V],
-  sourceDescription: List[String]
-) { self =>
-  final def orElse(that: => ConfigSource[K, V]): ConfigSource[K, V] =
-    ConfigSource(
-      k => getConfigValue(k).getOrElse(that.getConfigValue(k)),
-      if (self.sourceDescription == that.sourceDescription) self.sourceDescription
-      else self.sourceDescription ++ that.sourceDescription
-    )
+sealed trait ConfigSource[K, V] { self =>
+  def sourceDescription: List[String]
+
+  def getConfigValue(path: List[K]): PropertyTree[K, V]
+
+  final def orElse(that: ConfigSource[K, V]): ConfigSource[K, V] = MergedConfigSource(self, that)
 
-  final def <>(that: => ConfigSource[K, V]): ConfigSource[K, V] =
-    self orElse that
+  final def <>(that: => ConfigSource[K, V]): ConfigSource[K, V] = self orElse that
+}
+
+final case class TreeConfigSource[K, V](tree: PropertyTree[K, V], sourceDescription: List[String])
+    extends ConfigSource[K, V] {
+  def getConfigValue(path: List[K]): PropertyTree[K, V] = tree.getPath(path)
+}
+final case class MergedConfigSource[K, V](left: ConfigSource[K, V], right: ConfigSource[K, V])
```

Comment:
It looks like this exists mainly to avoid accumulating duplicate source descriptions.

I'd suggest that `sourceDescription` should really be `Set[String]` to ensure there can be no duplicates, during composition or otherwise.

Reply:
done